python/tests/test_factors.py

```diff
@@ -34,3 +34,63 @@ def test_revolute_joint_measurement_factor(self):
                                             self.robot.joint("joint_1"), np.pi / 4, self.k)
 
         self.assertTrue(isinstance(factor, gtd.JointMeasurementFactor))
+
+class TestDynamicsFactors(unittest.TestCase):
+    def setUp(self):
+        self.k = 0
+        ROBOT_FILE = osp.join(gtd.SDF_PATH, "test", "simple_rr.sdf")
+        self.robot = gtd.CreateRobotFromFile(str(ROBOT_FILE), "simple_rr_sdf")
+        self.link0 = self.robot.link("link_0")
+        self.link1 = self.robot.link("link_1")
+        self.joint1 = self.robot.joint("joint_1")
+        self.noise6 = gtsam.noiseModel.Isotropic.Sigma(6, 0.1)
+        self.noise3 = gtsam.noiseModel.Isotropic.Sigma(3, 0.1)
+        self.noise1 = gtsam.noiseModel.Isotropic.Sigma(1, 0.1)
+    
+        self.values = gtsam.Values()
+        gtd.InsertJointAngle(self.values, 0, self.k, 0.0)
+        gtd.InsertJointAngle(self.values, 1, self.k, 0.0)
+        self.values = self.robot.forwardKinematics(self.values, self.k, "link_0")
+        gtd.InsertTorque(self.values, 0, self.k, 0.0)
+        gtd.InsertWrench(self.values, 0, 0, self.k, np.zeros(6))
+        gtd.InsertWrench(self.values, 1, 0, self.k, np.zeros(6))
+
+    def test_pose_factor(self):
```

Comment:
These should ideally be their own class for each factor so we can have methods for construction, error evaluation and jacobian computation.

Reply:
I already started working on that, so you can use this as a template. Note that I want the constructor tests in particular since that exposes a lot of bugginess in the wrapping process.

```python
"""
 * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
 * Atlanta, Georgia 30332-0415
 * All Rights Reserved
 * See LICENSE for the license information
 *
 * @file  test_dynamics_graph.py
 * @brief Unit tests for dynamics graph.
 * @author Varun Agrawal
"""

import os.path as osp
import unittest

import gtdynamics as gtd
import gtsam
import numpy as np


class TestFactors(unittest.TestCase):
    """
    Base class for testing various factors.
    Provides needed fixtures and common functions.
    """
    def setUp(self):
        self.k = 0
        self.wTp_key = gtd.internal.PoseKey(0, self.k).key()
        self.wTc_key = gtd.internal.PoseKey(1, self.k).key()

        ROBOT_FILE = osp.join(gtd.SDF_PATH, "test", "simple_rr.sdf")
        self.robot = gtd.CreateRobotFromFile(str(ROBOT_FILE), "simple_rr_sdf")


class TestJointMeasurementFactor(TestFactors):
    """Test suite for various versions of the JointMeasurementFactor."""
    def test_revolute_joint_measurement_factor(self):
        """Test RevoluteJointMeasurementFactor."""
        factor = gtd.JointMeasurementFactor(
            gtsam.noiseModel.Isotropic.Sigma(6, 0.1),
            self.robot.joint("joint_1"), np.pi / 4, self.k)

        self.assertIsInstance(factor, gtd.JointMeasurementFactor)


class TestTempPoseFactor(TestFactors):
    """Test suite for various versions of the JointMeasurementFactor."""
    def test_temp_pose_factor_constructor(self):
        """Test constructor."""
        joint = self.robot.joint("joint_1")
        joint_key = gtd.internal.JointAngleKey(joint.id(), self.k).key()
        pose_model = gtsam.noiseModel.Isotropic.Sigma(6, 0.1)
        pose_factor = gtd.TempPoseFactor(self.wTp_key, self.wTc_key, joint_key,
                                         pose_model, joint)

        self.assertIsInstance(pose_factor, gtd.TempPoseFactor)

    def test_pose_factor_constructor(self):
        """Test constructor."""
        joint = self.robot.joint("joint_1")
        joint_key = gtd.internal.JointAngleKey(joint.id(), self.k).key()
        pose_model = gtsam.noiseModel.Isotropic.Sigma(6, 0.1)
        pose_factor = gtd.PoseFactor(self.wTp_key, self.wTc_key, joint_key,
                                     pose_model, joint)

        self.assertIsInstance(pose_factor, gtd.PoseFactor)


class TestContactKinematicsAccelFactor(TestFactors):
    def test_constructor(self):
        pass


class TestMinTorqueFactor(TestFactors):
    def test_constructor(self):
        pass


class TestPreintegratedContactFactors(TestFactors):
    def test_constructor(self):
        pass


class TestWrenchFactor(TestFactors):
    def test_constructor(self):
        pass


class TestCollocationFactors(TestFactors):
    def test_constructor(self):
        pass


class TestContactKinematicsTwistFactor(TestFactors):
    def test_constructor(self):
        pass


class TestTorqueFactor(TestFactors):
    def test_constructor(self):
        pass


class TestWrenchPlanarFactor(TestFactors):
    def test_constructor(self):
        pass


class TestContactDynamicsFrictionConeFactor(TestFactors):
    def test_constructor(self):
        pass


class TestContactPointFactor(TestFactors):
    def test_constructor(self):
        pass


class TestObjectiveFactors(TestFactors):
    def test_constructor(self):
        pass


class TestTwistAccelFactor(TestFactors):
    def test_constructor(self):
        pass


class TestContactDynamicsMomentFactor(TestFactors):
    def test_constructor(self):
        pass


class TestForwardKinematicsFactor(TestFactors):
    def test_constructor(self):
        pass


class TestTwistFactor(TestFactors):
    def test_constructor(self):
        pass


class TestContactEqualityFactor(TestFactors):
    def test_constructor(self):
        pass


class TestJointLimitFactor(TestFactors):
    def test_constructor(self):
        pass


class TestPointGoalFactor(TestFactors):
    def test_constructor(self):
        pass


class TestWrenchEquivalenceFactor(TestFactors):
    def test_constructor(self):
        pass


class TestContactHeightFactor(TestFactors):
    def test_constructor(self):
        pass

```

Reply:
temporarily remove all the factor tests and wrapper, except for the WrenchFactor to make cablerobot tests working